File: 2019/Practice/hard.py
MOD = 1000000007

def read_int():
    return int(input())
    
def read_ints():
    return list(map(int, input().split()))
    
def load():
    N, K, x1, y1, C, D, E1, E2, F = read_ints()
    A = get_array(N, x1, y1, C, D, E1, E2, F)   # O(N)
    
    return N, K, A
    
def solve(case):
    N, K, A = case
    
    # POWER = SUM(POWER_i) = SUM(digit_sums[i-1] * (i^1 + i^2 + ... + i^K))
    # which can be simplified by the sum formula of equal-ratio series:
    # i^1 + ... + i^K = i*(i^K-1)/(i-1)
    
    digit_sum = 0
    power = 0
    for i in range(N):
        digit_sum = (digit_sum + A[N-1-i] * (i+1)) % MOD

        if i == N - 1:
            power = (power + digit_sum * K) % MOD
        else:
            power = (power + digit_sum * (N-i) * (pow(N-i, K, MOD) - 1) * pow(N-i-1, MOD-2, MOD)) % MOD
    
    return power
    
def get_array(N, x1, y1, C, D, E1, E2, F):
    X = [x1 % F]
    Y = [y1 % F]
    A = [(x1 + y1) % F]
    C %= F
    D %= F
    E1 %= F
    E2 %= F
    
    for i in range(1, N):
        x_next = (C * X[-1] + D * Y[-1] + E1) % F
        y_next = (D * X[-1] + C * Y[-1] + E2) % F
        X.append(x_next)
        Y.append(y_next)
        A.append((x_next + y_next) % F)
    
    return A

def get_sum(A):
    L = len(A)
    
    # denote f(i, j) as the total times appeared for A[i] in all contiguous subarrays at position j
    # therefore f(i, j) = 0, for i < j
    # and f(i, j) = L - i, otherwise
    
    # note that POWER_x is the summation of power of all contiguous subarrays
    # and power of each contiguous subarray is calculated in digit-wise fashion
    # so we can sum up each digits first,
    # and then apply the exponential multiplicants
    
    # sum of digit position j for all contiguous subarrays:
    # SUM(f(i, j) * A[i] for i >= j)
    
    # calculate in reverse order, to reuse the former result
    
    digit_sums = []
    s = 0
    for j in range(L-1, -1, -1):
        s += A[j] * (L - j)
        digit_sums.append(s)
    
    return list(reversed(digit_sums))

if __name__ == '__main__':
    T = read_int()
    
    for i in range(T):
        print('Case #{}: {}'.format(i+1, solve(load())))

File: 2019/Practice/hard.cpp
#include <iostream>
#include <vector>
using namespace std;
typedef unsigned long long ull;

ull mod = 1000000007;

struct Case {
    ull N;
    ull K;
    vector<ull> A;
    vector<ull> digit_sums;
};

vector<ull> get_array(ull N, ull x1, ull y1, ull C, ull D, ull E1, ull E2, ull F) {
    x1 %= F, y1 %= F;
    C %= F, D %= F;
    E1 %= F, E2 %= F;
    
    vector<ull> X = {x1};
    vector<ull> Y = {y1};
    vector<ull> A = {(x1 + y1) % F};
    
    ull x, y;
    ull xb = x1, yb = y1;
    for (ull i = 0; i < N; ++i) {
        x = (C * xb + D * yb + E1) % F;
        y = (D * xb + C * yb + E2) % F;
        X.push_back(x);
        Y.push_back(y);
        A.push_back((x + y) % F);
        
        xb = x, yb = y;
    }
    
    return A;
}

vector<ull> get_sum(const vector<ull>& A, ull size) {
    vector<ull> digit_sums (size, 0);
    ull s = 0;
    
    for (ull i = 0; i < size; ++i) {
        s = (s + A[size-1-i] * (i + 1)) % mod;
        digit_sums[size-1-i] = s;
    }
    
    return digit_sums;
}

Case load() {
    Case cs;
    ull x1, y1;
    ull C, D, E1, E2, F;
    cin >> cs.N >> cs.K >> x1 >> y1 >> C >> D >> E1 >> E2 >> F;
    
    cs.A = get_array(cs.N, x1, y1, C, D, E1, E2, F);
    cs.digit_sums = get_sum(cs.A, cs.N);
    
    return cs;
}

ull fast_pow(ull a, ull b, ull c=1000000007) {

    if (a == 1) {
        return a;
    }

    ull res = 1;
    
    while (b > 0) {
        if (b & 0x01) {
            res = (res * a) % c;
        }

        b >>= 1;
        a = (a * a) % c;
    }

    return res;
}

ull solve(Case cs) {
    ull power = (cs.digit_sums[0] * cs.K) % mod;
    ull t;

    for (ull i = 1; i < cs.N; ++i) {
        t = (cs.digit_sums[i] * (i+1)) % mod;
        t = (t * (fast_pow(i+1, cs.K) - 1)) % mod;
        t = (t * fast_pow(i, mod-2)) % mod;

        power += t;
        power %= mod;
    }
    
    return power;
}

int main() {
    int T;
    cin >> T;
    
    for (int i = 0; i < T; ++i) {
        cout << "Case #" << i+1 << ": " << solve(load()) << "\n";
    }
    
    return 0;
}